src/Types/ListType.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace ILIAS\Types;
+
+/**
+ * Class ListType
+ *
+ * @package ILIAS\Types
+ *
+ * @author  Oskar Truffer <ot@studer-raimann.ch>
+ *
+ * Describes a list of a certain Type.
+ *
+ * Attention:
+ * isExtensionOf behaves Covariant!
+ * new ListType(Dog::class).isExtensionOf(new ListType(Animal::class)) == true!
+ *
+ * See: https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)
+ */
+class ListType implements Type, Ancestors {
+
+	/** @var Type */
+	protected $type;
+
+
+	/**
+	 * SingleType constructor.
+	 *
+	 * @param $fullyQualifiedClassName string|Type Give a Value Type or a Type that will be wrapped in a single type.
+	 */
+	public function __construct($fullyQualifiedClassName) {
+		if (!is_a($fullyQualifiedClassName, Type::class)) {
+			$fullyQualifiedClassName = new SingleType($fullyQualifiedClassName);
+		}
+		$this->type = $fullyQualifiedClassName;
+	}
+
+
+	/**
+	 * @inheritdoc
+	 */
+	function __toString() {
+		return "[" . $this->type . "]";
+	}
+
+
+	/**
+	 * Is this type a subtype of $type. Not strict! x->isExtensionOf(x) == true.
+	 *
+	 * Attention: [Dog].isExtensionOf([Animal]) == true. In other words:
+	 * The isExtensionOf behaves Covariant!
+	 *
+	 * If you are familiar with e.g. Java Generics ?.isExtensionOf(x) behaves the same as <? extends x>.
```

Comment:
Hi Oskar!

I'm happy we sorted that one out =) Indeed, List<? extends B> is not sound as well.

Possible solutions to the problem seem to be:
* Split covariant and contravariant part of the list. Go on as you proposed with these two separated. I think for your usecase the `pop`-part could be sufficient. Could be hard to find names, though, and won't help if arrays are in fact passed.
* Drop this whole notion of `extension` or `subclass`. AFAIK this is the source of the trouble. You want to able to stuff a list of courses in a place where a list of objects is processed? Add an explicit transformation step.
* Keep this whole notion of `extension` or `subclass`. Don't lie about it for the list. You want to able to stuff a list of courses in a place where a list of objects is processed? Add an explicit transformation step.

I think I would go with the second one. I somehow doubt, that all inheritance-hierarchies in ILIAS respect LSP, so the last approach will be doomed anyway. In general, "implementing some interface" to me seems to be a saner concept than "inheriting from some other class". The former seems to say more about expected behaviour, while the latter seems to be mostly used as a way to share code (in general, not only in ILIAS). There might be some solution for your course/object-case hidden in the interface approach.

Best regards!